File: src/nostr/rpc_machine.rs
use crate::nostr::{Event, Filter, RelayMessage};

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum RpcState {
    Initial,
    AwaitingResponse,
    Success(Event),
    Failed(String),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum RpcAction {
    Subscribe(String, Filter),
    Publish(Event),
    Unsubscribe(String),
}

pub struct NwcRpcMachine {
    request: Event,
    state: RpcState,
    sub_id: String,
}

impl NwcRpcMachine {
    pub fn new(request: Event) -> Self {
        Self {
            request,
            state: RpcState::Initial,
            sub_id: "rpc_sub".to_string(),
        }
    }

    pub fn state(&self) -> &RpcState {
        &self.state
    }

    pub fn start(&mut self) -> Vec<RpcAction> {
        let filter = Filter {
            e_tags: Some(vec![self.request.id.clone()]),
            p_tags: Some(vec![self.request.pubkey.clone()]),
            ..Filter::default()
        };

        self.state = RpcState::AwaitingResponse;

        vec![
            RpcAction::Subscribe(self.sub_id.clone(), filter),
            RpcAction::Publish(self.request.clone()),
        ]
    }

    pub fn transition(&mut self, message: RelayMessage) -> Option<RpcAction> {
        match (&self.state, message) {
            (RpcState::AwaitingResponse, RelayMessage::Event(_, event)) => {
                // Correlation check: ensure this event references our request
                // In NWC, response should have an 'e' tag pointing to request ID
                let references_request = event.tags.iter().any(|t| {
                    matches!(t, crate::nostr::Tag::E(eid, _) if eid == &self.request.id)
                });

                if references_request {
                    self.state = RpcState::Success(event);
                    Some(RpcAction::Unsubscribe(self.sub_id.clone()))
                } else {
                    None
                }
            }
            (RpcState::AwaitingResponse, RelayMessage::Eose(_)) => {
                // Protocol noise, stay in current state
                None
            }
            (RpcState::AwaitingResponse, RelayMessage::Notice(msg)) => {
                self.state = RpcState::Failed(format!("Relay notice: {}", msg));
                Some(RpcAction::Unsubscribe(self.sub_id.clone()))
            }
            _ => None,
        }
    }

}

#[cfg(test)]
#[path = "rpc_machine_test.rs"]
mod rpc_machine_test;

File: src/nostr/rpc-machine.ts
import type { Event } from './event.ts';
import type { Filter } from './filter.ts';
import type { RelayMessage } from './nip01.ts';

export type RpcState =
  | { kind: 'initial' }
  | { kind: 'awaitingResponse' }
  | { kind: 'success'; event: Event }
  | { kind: 'failed'; reason: string };

export type RpcAction =
  | { kind: 'subscribe'; subId: string; filter: Filter }
  | { kind: 'publish'; event: Event }
  | { kind: 'unsubscribe'; subId: string };

export class NwcRpcMachine {
  private request: Event;
  private state: RpcState = { kind: 'initial' };
  private subId = 'rpc_sub';

  constructor(request: Event) {
    this.request = request;
  }

  getState(): RpcState {
    return this.state;
  }

  start(): RpcAction[] {
    const filter: Filter = {
      eTags: [this.request.id],
      pTags: [this.request.pubkey],
    };

    this.state = { kind: 'awaitingResponse' };

    return [
      { kind: 'subscribe', subId: this.subId, filter },
      { kind: 'publish', event: this.request },
    ];
  }

  transition(message: RelayMessage): RpcAction | null {
    if (this.state.kind !== 'awaitingResponse') return null;

    switch (message.type) {
      case 'EVENT': {
        const referencesRequest = message.event.tags.some(
          t => t.eventId() === this.request.id,
        );
        if (referencesRequest) {
          this.state = { kind: 'success', event: message.event };
          return { kind: 'unsubscribe', subId: this.subId };
        }
        return null;
      }
      case 'EOSE':
        return null;
      case 'NOTICE':
        this.state = { kind: 'failed', reason: `Relay notice: ${message.message}` };
        return { kind: 'unsubscribe', subId: this.subId };
      default:
        return null;
    }
  }
}
